Add logistic regression tutorial

Ishitori opened the pull request:
## Description ##
Add new tutorial on how to do logistic regression with Gluon API (writing minimal amount of custom code) + provided some explanation why it should be done in that way

## Checklist ##
### Essentials ###
- [ ] Changes are complete (i.e. I finished coding on this PR)
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [ ] Code is well-documented: 
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] New tutorial on logistic regression

Ishitori commented:
@thomelane @ThomasDelteil @safrooze or @indhub please take a look

ifeherva commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+Loss function is used to calculate how the output of the network different from the ground truth. In case of the logistic regression the ground truth are class labels, which can be either 0 or 1. Because of that, we are using `SigmoidBinaryCrossEntropyLoss`, which suites well for that scenario.
+
+Trainer object allows to specify the method of training to be used. There are various methods available, and for our tutorial we use a widely accepted method Stochastic Gradient Descent. We also need to parametrize it with learning rate value, which defines how fast training happens, and weight decay which is used for regularization.
```
"We also need to parametrize it with learning rate value, which defines how fast training happens" The learning rate of SGD defines the weight updates, not necessarily how fast training happens. I propose to reword it a bit to avoid confusion that large LR will lead to fast training.

Ishitori replied:
Good suggestion, fixed.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+Loss function is used to calculate how the output of the network different from the ground truth. In case of the logistic regression the ground truth are class labels, which can be either 0 or 1. Because of that, we are using `SigmoidBinaryCrossEntropyLoss`, which suites well for that scenario.
+
+Trainer object allows to specify the method of training to be used. There are various methods available, and for our tutorial we use a widely accepted method Stochastic Gradient Descent. We also need to parametrize it with learning rate value, which defines how fast training happens, and weight decay which is used for regularization.
```
Suggest to make a brief statement that you're using SGD, and if needed, include a link for further reading.

Ishitori replied:
I mention Stochastic Gradient Descent in the text above. So, made the text shorter and added links to the docs and to the Straight Dope tutorial.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+After defining the model, we need to define a few more thing: our loss, our trainer and our metric.
+
+Loss function is used to calculate how the output of the network different from the ground truth. In case of the logistic regression the ground truth are class labels, which can be either 0 or 1. Because of that, we are using `SigmoidBinaryCrossEntropyLoss`, which suites well for that scenario.
```
I think if instead of defining what a loss is, you'd define what `SigmoidBinaryCrossEntropyLoss` does, user may benefit more. Something that explains there is a `from_sigmoid` argument, which means that the output of the neuron doesn't need to go through sigmoid, but at inference you'd have to pass it through sigmoid. Also include link to API.

Ishitori replied:
I decided to not remove a sentence explaining what loss does, but add links to official documentation and wikipedia. I also pasted your explanation of from_sigmoid - it might be useful to explain it even when we don't set this argument.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
````diff
@@ -0,0 +1,215 @@
+accuracy = mx.metric.Accuracy()
+```
+
+Usually, it is not enough to pass the training data through a network only once to achieve high Accuracy. It helps when the network sees each example multiple times. The number of displaying every example to the network is called `epoch`. How big this number should be is unknown in advance, and usually it is estimated using trial and error approach. 
````
Suggest to completely skip description about what an epoch is.

Ishitori replied:
Sure, not much value here.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+Usually, it is not enough to pass the training data through a network only once to achieve high Accuracy. It helps when the network sees each example multiple times. The number of displaying every example to the network is called `epoch`. How big this number should be is unknown in advance, and usually it is estimated using trial and error approach. 
+
+Below we are defining the main training loop, which go through each example in batches specified number of times (epochs). After each epoch we display training loss, validation loss and calculate accuracy of the model using validation set. For now, let's take a look into the code, and I am going to explain the details later.
```
"validataion dateset"

Ishitori replied:
Cannot find this typo.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+        # Calculate gradients 
+        loss_result.backward()
+
+        # Change parameters of the network
```
use "update" instead of "change"

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+    print("Epoch: %s, Training loss: %.2f, Validation loss: %.2f, Validation accuracy: %s" % 
+          (e, cumulative_train_loss, cumulative_val_loss, accuracy.get()[1]))
+
+    # we reset accuracy, so the new epoch's accuracy would be calculate from the blank state
```
would be ***calculated***

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+        # we sum losses of every batch per an epoch into a single variable
+        cumulative_train_loss += nd.sum(loss_result).asscalar()
+
+    # iterate over all batches of validation data and calculate validation loss
```
Split into train and validate function.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+        # Similar to cumulative training loss, calculate cumulative validation loss
+        cumulative_val_loss += nd.sum(loss(output, val_ground_truth_class)).asscalar()
+        
+        # Applying sigmoid function, to get data in range [0, 1] and then
```
In a notebook, better to keep long comments as text, rather than comment section of code.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+        # and update accuracy with the results for that batch.
+        accuracy.update(val_ground_truth_class, prediction.reshape(-1)) 
+    
+    # in the end of epoch, we print out current values for epoch, training and validation losses, and accuracy
```
Stating the obvious.

Ishitori replied:
Removed the print comment, simplified the `accuracy.update` comment

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+        # Do forward pass on a batch of validation data
+        output = net(val_data)
+        
+        # Similar to cumulative training loss, calculate cumulative validation loss
```
stating the obvious

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+    # iterate over all batches of validation data and calculate validation loss
+    for i, (val_data, val_ground_truth_class) in enumerate(val_dataloader):
+        # Do forward pass on a batch of validation data
```
stating the obvious.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+    # iterate over all batches of training data and calculate training loss
+    for i, (data, label) in enumerate(train_dataloader):
+        with autograd.record():
+            # Do forward pass on a batch of training data
```
stating the obvious.

Ishitori replied:
Removed the loop comment, but kept the forward pass one.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+    cumulative_train_loss = 0
+    cumulative_val_loss = 0
+    
+    # iterate over all batches of training data and calculate training loss
```
stating the obvious.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+epochs = 10
+threshold = 0.5
+
+# iterate over training data epochs number of times
```
stating the obvious.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+## Tip 2: Encode classes as 0 and 1
+
+`Sigmoid` function produces values from 0 to 1. `SigmoidBinaryCrossEntropyLoss` uses these values to calculate the   loss by essentially subtracting values and class labels from 1. [Here is the formula](https://mxnet.incubator.apache.org/api/python/gluon/loss.html?highlight=sigmoidbinarycrossentropyloss#mxnet.gluon.loss.SigmoidBinaryCrossEntropyLoss) used for that calculation (we use default vers ...
```
Don't get the last sentence. What is exactly better?

Ishitori replied:
Rewrote whole paragraph simplifying.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+`Sigmoid` function produces values from 0 to 1. `SigmoidBinaryCrossEntropyLoss` uses these values to calculate the   loss by essentially subtracting values and class labels from 1. [Here is the formula](https://mxnet.incubator.apache.org/api/python/gluon/loss.html?highlight=sigmoidbinarycrossentropyloss#mxnet.gluon.loss.SigmoidBinaryCrossEntropyLoss) used for that calculation (we use default vers ...
+
+If your data comes with a label encoded in a different format, such as -1 and 1, then you can either recode it to 0 and 1 by comparing the initial class to 0, or use another function instead of `Sigmoid`, like [`Tanh`](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html?highlight=tanh#mxnet.ndarray.tanh), to produce output in range [-1; 1].
```
Not sure if mentioning tanh has any merit here. tanh is just not used for logistic regression. What loss do I use with tanh? I find it confusing.

Ishitori replied:
Removed Tanh part.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+If your data comes with a label encoded in a different format, such as -1 and 1, then you can either recode it to 0 and 1 by comparing the initial class to 0, or use another function instead of `Sigmoid`, like [`Tanh`](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html?highlight=tanh#mxnet.ndarray.tanh), to produce output in range [-1; 1].
+
+## Tip 3: Use SigmoidBinaryCrossEntropyLoss instead of LogisticRegressionOutput
```
Include link to API.

Ishitori replied:
This block was deleted.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+NDArray API has two options to calculate logistic regression loss. One is `SigmoidBinaryCrossEntropyLoss`, which I used in the example. This class inherits from the `Loss` class and is intended to be used as a loss function for logistic regression. But there is also a function called `LogisticRegressionOutput`, which can be applied to any `NDArray`. Mathematically speaking, this function does the ...
+
+My recommendation would be to use `SigmoidBinaryCrossEntropyLoss`, because this class properly inherits from `Loss` class, while `LogisticRegressionOutput` is just a regular function. `LogisticRegressionOutput` is a   function to go when implementing logistic regression using Symbol API, but in case of using Gluon API, there are no benefits using it. The only case when you may want to consider us ...
```
This paragraph needs rewording. It doesn't really make a case for why not use LogisticRegressionOutput (i.e. the fact that it is designed to be an output layer when using the Module API). Also what's the problem with using sparse matrices and SigmoidBinaryCrossEntropyLoss?! What happens if I do? What fails? You may consider just getting rid of this tip altogether to avoid any confusion.

Ishitori replied:
If you try to use `SigmoidBinaryCrossEntropyLoss` with sparse data, it would throw an exception at some point, where it tries to do a reshape:

```NotSupportedForSparseNDArray: Function reshape with arguments (<class 'tuple'>) is not supported for SparseNDArray and only available in NDArray.```

Ishitori replied:
I rewrote the paragraph, and I actually deleted the sparse part from it as well. @eric-haibin-lin is going to work on this at some point, so to not update the tutorial later, I just omit this part.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+## Tip 4: Convert probabilities to classes before calculating Accuracy
+
+`Accuracy` metric requires 2 arguments: 1) a vector of ground-truth classes and 2) A tensor of predictions. When tensor of predictions is of the same shape as the vector of ground-truth classes, `Accuracy` class assumes that it should contain predicted classes. So, it converts the vector to `Int32` and compare each item of ground-truth classes to prediction vector. 
```
Why is one called a vector and the other called a tensor?

Ishitori replied:
Replaced it with "vector of matrix"

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+`Accuracy` metric requires 2 arguments: 1) a vector of ground-truth classes and 2) A tensor of predictions. When tensor of predictions is of the same shape as the vector of ground-truth classes, `Accuracy` class assumes that it should contain predicted classes. So, it converts the vector to `Int32` and compare each item of ground-truth classes to prediction vector. 
+
+Because of the behaviour above, you will get an unexpected result if you just pass the output of `Sigmoid` function as is. `Sigmoid` function produces output in range [0; 1], and all numbers in that range are going to be casted to 0, even if it is as high as 0.99. To avoid this we write a custom bit of code, that:
```
Instead of having a tip here, I recommend to write the rounding as a separate function in notebook and include a paragraph on top explaining what it's doing. This paragraph is really the only motivation for writing this tutorial.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+The same is not true, if the output shape of your function is different from the shape of ground-truth classes vector. For example, when doing multiclass regression with `Softmax` as an output, the shape of the output is going to be *number_of_examples* x *number_of_classes*. In that case we don't need to do the transformation above, because `Accuracy` metric would understand that shapes are diff ...
+
+This makes things a little bit easier, and that's why I have seen examples where `Softmax` is used as an output of prediction. If you want to do that, make sure to change the output layer size to 2 neurons, where each neuron will provide a value of an example to belong to class 0 and 1 respectively.
```
Wouldn't call it *easier*. Softmax allows the output layer to have more flexibility in learning because it contains twice the number of parameters to learn.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+## Conclusion
+
+In this tutorial I explained some potential pitfalls to be aware about when doing logistic regression in Apache MXNet Gluon API. There might be some other challenging scenarios, which are not covered in this tutorial, like dealing with imbalanced classes, but I hope this tutorial will serve as a guidance and all other potential pitfalls would be covered in future tutorials.
```
"be aware of"

Ishitori replied:
Fixed

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+## Conclusion
+
+In this tutorial I explained some potential pitfalls to be aware about when doing logistic regression in Apache MXNet Gluon API. There might be some other challenging scenarios, which are not covered in this tutorial, like dealing with imbalanced classes, but I hope this tutorial will serve as a guidance and all other potential pitfalls would be covered in future tutorials.
```
Suggest to keep the conclusion factual and not include reference to one of many challenges or "hoping" the tutorial is useful. Also this is not a blog (no posting time associated with it). Future tutorials has no meaning.

Ishitori replied:
Rewrote the conclusion by reiterating over main points of the tutorial.

Ishitori commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+# Logistic regression using Gluon API explained
```
There is a lot of similarity with the tutorial you have mentioned. My personal pet-peeve with the tutorial in The Straight Dope is that it doesn't use library functions for sigmoid, loss and accuracy. I believe it creates an impression that you need to write a lot of boiler plate code instead of using what MXNet offers. 

Also, there were confusions around number of neurons in forum and stackoverflow which is directly relates to the usage of `Accuracy`, but needs a bit more context to cover.

Ishitori commented on docs/tutorials/gluon/logistic_regression_explained.md:
````diff
@@ -0,0 +1,215 @@
+mx.random.seed(12345)  # Added for reproducibility
+```
+
+In this tutorial we will use fake dataset, which contains 10 features drawn from a normal distribution with mean equals to 0 and standard deviation equals to 1, and a class label, which can be either 0 or 1. The length of the dataset is an arbitrary value. The function below helps us to generate a dataset.
````
I didn't want to add extra code for data loading and data processing, because this is not the point of this tutorial. The optimal way would be if there is a binary classification dataset in the mxnet itself, so it can be loaded in one line and no pre-processing would be required.

Ishitori commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+## Working with data
+
+To work with data, Apache MXNet provides Dataset and DataLoader classes. The former is used to provide an indexed access to the data, the latter is used to shuffle and batchify the data. 
```
Yes, added links and remove my explanation. Added a link to Datasets and Dataloaders tutorial instead.

Ishitori commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+This separation is done because a source of Dataset can vary from a simple array of numbers to complex data structures like text and images. DataLoader doesn't need to be aware of the source of data as long as Dataset provides a way to get the number of records and to load a record by index. As an outcome, Dataset doesn't need to hold in memory all data at once. Needless to say, that one can impl ...
+
+Below we define 2 datasets: training dataset and validation dataset. It is a good practice to measure performance of a trained model on a data that the network hasn't seen before. That is why we are going to use training set for training the model and validation set to calculate model's accuracy.
```
Agree. Removed this one as well.

safrooze replied:
"training set" -> "training dataset"
"validation set" -> "validation dataset"

Ishitori commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+## Defining and training the model
+
+In real application, model can be arbitrary complex. The only requirement for the logistic regression is that the last layer of the network must be a single neuron. Apache MXNet allows us to do so by using `Dense` layer and specifying the number of units to 1.
```
Done

Ishitori commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -0,0 +1,215 @@
+
+In real application, model can be arbitrary complex. The only requirement for the logistic regression is that the last layer of the network must be a single neuron. Apache MXNet allows us to do so by using `Dense` layer and specifying the number of units to 1.
+
+Below, we define a model which has an input layer of 10 neurons, a couple inner layers of 10 neurons each, and output layer of 1 neuron, as it is required by logistic regression. We stack the layers using `HybridSequential` block and initialize parameters of the network using `Xavier` initialization. 
```
Fixed

Ishitori commented on docs/tutorials/gluon/logistic_regression_explained.md:
````diff
@@ -0,0 +1,215 @@
+net.initialize(mx.init.Xavier(magnitude=2.34))
+```
+
+After defining the model, we need to define a few more thing: our loss, our trainer and our metric.
````
Fixed

yifeim commented:
The tutorial looks awesome! A few comments:

* Is it just me or everybody experiences training loss 8x larger than validation loss? I understand it may be possible due to unstable weights from backward update after every batch. But, still it is not my usual expectation. (Not to say that I have not seen similar issues before and I am still slightly puzzled by the root cause.)

* The model class is linear, but the proposed network is a 3-layer overkill (which may be related to the larger training loss). If you want a fun problem, maybe consider a `xor` function class: https://medium.com/@jayeshbahire/the-xor-problem-in-neural-networks-50006411840b

* Is this `mxnet.gluon.loss.LogisticLoss` equivalent to `SigmoidBCELoss`? Since you are using metrics, it may be worthwhile exploring `mx.metric.F1`.

* There are some magic hyperparameters to be explained: a strong wd=0.01, an Xavier=2.34 initialization. As an elementary tutorial, I would try to simplify them unless they are part of the intended purposes.

safrooze commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -120,96 +108,115 @@ for e in range(epochs):
+Despite that there are 2 classes, there should be only one output neuron, because `SigmoidBinaryCrossEntropyLoss` accepts only one feature as an input. 
 
-1. Subtracts a threshold from the original prediction. Usually, the threshold is equal to 0.5, but it can be higher, if you want to increase certainty of an item to belong to class 1.
+In case when there are 3 or more classes, one cannot use a single Logistic regression, but should do multiclass regression. The solution would be to increase the number of output neurons to the number of classes and use `SoftmaxCrossEntropyLoss`. 
```
* "In case there are".
* Multi-class regression refers to the case when one item belongs to multiple classes. In that case you cannot use softmax, but rather multiple logistic regressions, one for each class. I suggest to simply remove this sentence to not create any confusion.

Ishitori replied:
Removed

Ishitori commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -59,9 +55,9 @@ val_dataloader = DataLoader(val_dataset, batch_size=batch_size, shuffle=True)
 ## Defining and training the model
 
-In real application, model can be arbitrary complex. The only requirement for the logistic regression is that the last layer of the network must be a single neuron. Apache MXNet allows us to do so by using `Dense` layer and specifying the number of units to 1.
+The only requirement for the logistic regression is that the last layer of the network must be a single neuron. Apache MXNet allows us to do so by using [Dense](https://mxnet.incubator.apache.org/api/python/gluon/nn.html#mxnet.gluon.nn.Dense) layer and specifying the number of units to 1. The rest of the network can be arbitrary complex. 
```
Fixed

Ishitori commented on docs/tutorials/gluon/logistic_regression_explained.md:
```diff
@@ -94,21 +90,13 @@ trainer = Trainer(params=net.collect_params(), optimizer='sgd',
-Usually, it is not enough to pass the training data through a network only once to achieve high Accuracy. It helps when the network sees each example multiple times. The number of displaying every example to the network is called `epoch`. How big this number should be is unknown in advance, and usually it is estimated using trial and error approach. 
-
-Below we are defining the main training loop, which go through each example in batches specified number of times (epochs). After each epoch we display training loss, validation loss and calculate accuracy of the model using validation set. For now, let's take a look into the code, and I am going to explain the details later.
+The next step is to define the training function in which we iterate over all batches of training data, execute the forward pass on each batch and calculate training loss. On line 19, we sum losses of every batch per an epoch into a single variable, because we calculate loss per single batch, but want to display it per epoch.
```
Fixed

Ishitori commented:
@yifeim, Thanks for comments, I am answering them in order:

1. The loss is so different, because I didn't do the averaging by number of items in training and validation set respectively. You are right that it is better to do so, otherwise numbers look confusing. Did that.

2. While XOR problem is fun and gives a good understanding why you need activation functions, it requires inputs of 0 and 1 as well. I wouldn't want to limit this part of the tutorial to this, as it seems a better use case to showcase arbitrary inputs.

3. `LogisticLoss` can be used instead of `SigmoidBCELoss`. There is a good explanation about both of them: https://stats.stackexchange.com/questions/229645/why-there-are-two-different-logistic-loss-formulation-notations/231994  

I haven't seen `LogisticLoss` used much in practice. I don't want to mention the second option, because it is going to confuse these who are new to Apache MXNet.

I added a showcase how to use F1 metric.

4. You are probably right - I didn't have an intention of going into default parameters as it may require long explanation. Just removed wd and use default of Xavier.

yifeim commented:
Awesome. Done and done! Thanks for addressing the questions.
